### PR TITLE
refactor: rewrite file_fetcher to use async fns, lift blocking call

### PR DESCRIPTION
**Pull request by bartlomieju**

### CODE REVIEW

File: cli/file_fetcher.rs [hunk 1]
```diff
@@ -213,54 +211,38 @@ impl SourceFileFetcher {
...
     if is_local_file {
-      match self.fetch_local_file(&module_url) {
-        Ok(source_file) => {
-          return Either::Left(futures::future::ok(source_file));
-        }
-        Err(err) => {
-          return Either::Left(futures::future::err(err));
-        }
-      }
+      return self.fetch_local_file(&module_url);
```
Comment (ry) [inline cli/file_fetcher.rs]: So much nicer.

File: cli/file_fetcher.rs [hunk 2]
```diff
@@ -213,54 +211,38 @@ impl SourceFileFetcher {
...
-        )
-        .into(),
-      ));
+      let e = std::io::Error::new(
+        std::io::ErrorKind::NotFound,
+        format!(
+          "Not allowed to get remote file '{}'",
+          module_url.to_string()
+        ),
+      );
```
Comment (ry) [inline cli/file_fetcher.rs]: Aside, not for this patch: basically anywhere we're constructing errors we should be using some predefined utility functions from cli/deno_error.rs. So this code would ideally be written:
```
deno_error::not_found(format!("Not allowed to get remote file '{}'", module_url.to_string())
```

Comment (bartlomieju) [reply cli/file_fetcher.rs]: Yes!!! Completely agree, this becomes very obvious working on #3936

File: cli/file_fetcher.rs [hunk 3]
```diff
@@ -127,79 +123,81 @@ impl SourceFileFetcher {
...
-
-    futures::executor::block_on(fut).ok()
+    // using "--cached-only" flag.
+    // It should be safe to for caller block on this
+    // future, because it doesn't actually do any asynchronous
+    // action in that path.
+    self
+      .get_source_file_async(specifier.as_url(), true, false, true)
+      .await
+      .ok()
```
Comment (ry) [inline cli/file_fetcher.rs]: It seems like if you're fetching a cached source file, you should be using sync ops. I see that it was using block_on before, so this is better. However, is there not a way to load this synchronously?

Comment (bartlomieju) [reply cli/file_fetcher.rs]: To be determined - most likely this will be completely removed, when I get to untangle TsCompiler from `file_fetcher`

Comment (ry) [reply cli/file_fetcher.rs]: Could you add a TODO here that fetching cache should use blocking fs syscalls?

Comment (bartlomieju) [reply cli/file_fetcher.rs]: Done

**Pull request merged**